plugins/events/eventrouter.go

```diff
@@ -0,0 +1,209 @@
+	"github.com/wavefronthq/wavefront-kubernetes-collector/internal/metrics"
+	"github.com/wavefronthq/wavefront-sdk-go/event"
+	v1 "k8s.io/api/core/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+var Log = log.WithField("system", "events")
+var leadershipName = "wf_collector_events"
+var filteredEvents = gometrics.GetOrRegisterCounter("events.filtered", gometrics.DefaultRegistry)
+var receviedEvents = gometrics.GetOrRegisterCounter("events.recevied", gometrics.DefaultRegistry)
+var sentEvents = gometrics.GetOrRegisterCounter("events.sent", gometrics.DefaultRegistry)
+
+type EventRouter struct {
+	kubeClient        kubernetes.Interface
+	eLister           corelisters.EventLister
+	eListerSynched    cache.InformerSynced
+	sink              metrics.DataSink
+	sharedInformers   informers.SharedInformerFactory
+	stop              chan struct{}
+	daemon            bool
+	leadershipManager *leadershipManager
+	whitelist         map[string]glob.Glob
+	blacklist         map[string]glob.Glob
+}
+
+func CreateEventRouter(clientset kubernetes.Interface, cfg *configuration.EventsConfig, sink metrics.DataSink, clusterName string, daemon bool) *EventRouter {
+	sharedInformers := informers.NewSharedInformerFactory(clientset, time.Minute)
+	eventsInformer := sharedInformers.Core().V1().Events()
+
+	er := &EventRouter{
+		kubeClient:      clientset,
+		sink:            sink,
+		daemon:          daemon,
+		sharedInformers: sharedInformers,
+		whitelist:       filter.MultiCompile(cfg.Filters.TagWhitelist),
+		blacklist:       filter.MultiCompile(cfg.Filters.TagBlacklist),
+	}
+
+	eventsInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: er.addEvent,
+	})
+	er.eLister = eventsInformer.Lister()
+	er.eListerSynched = eventsInformer.Informer().HasSynced
+
+	if er.daemon {
+		er.leadershipManager = newLeadershipManager(er, leadershipName, clientset)
+	}
+	return er
+}
+
+func (er *EventRouter) Start() {
+	if er.daemon {
+		er.leadershipManager.Start()
+	} else {
+		go func() { er.resume() }()
+	}
+}
+
+func (er *EventRouter) resume() {
+	er.stop = make(chan struct{})
+	defer utilruntime.HandleCrash()
+
+	Log.Infof("Starting EventRouter")
+
+	go func() { er.sharedInformers.Start(er.stop) }()
+
+	// here is where we kick the caches into gear
+	if !cache.WaitForCacheSync(er.stop, er.eListerSynched) {
+		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+		return
+	}
+	<-er.stop
+
+	Log.Infof("Shutting down EventRouter")
+}
+
+func (er *EventRouter) pause() {
+	close(er.stop)
+}
+
+func (er *EventRouter) Stop() {
+	if er.daemon {
+		er.leadershipManager.Stop()
+	}
+	er.pause()
+}
+
+// addEvent is called when an event is created, or during the initial list
+func (er *EventRouter) addEvent(obj interface{}) {
+	e := obj.(*v1.Event)
```

Comment:
Unlikely but this will panic if obj is not an event. Let's get the boolean and simply return for that case.